src/cryptokit.ml

```diff
@@ -2072,6 +2074,98 @@ let new_key ?(rng = Random.secure_rng) ?e numbits =
 
 end
 
+module Paillier = struct
+
+  type key =
+    { size: int;
+      n: string;
+      n2: string;
+      g: string;
+      p: string;
+      q: string;
+      lambda: string;
+      mu: string}
+  let wipe_key k =
+    wipe_string k.n;
+    wipe_string k.n2;
+    wipe_string k.g;
+    wipe_string k.p;
+    wipe_string k.q;
+    wipe_string k.lambda;
+    wipe_string k.mu
+
+  let encrypt key msg ?r ()=
+    let check_random_number r =
+      if (Bn.(relative_prime r (Bn.of_bytes key.n)) && r<(Bn.of_bytes key.n)) then true else false in
+    let rec get_r () =
+      let r = Bn.random ~rng:(Random.secure_rng#random_bytes) (key.size-1) in
+      if check_random_number r then r else get_r () in
+    let r = match r with
+      | None -> get_r ()
+      | Some r -> if check_random_number r then r else raise(Error Wrong_random_number )in
+    let msg = Bn.of_bytes msg in
+    let n = Bn.of_bytes key.n in
+    let n2 = Bn.of_bytes key.n2 in
+    let g = Bn.of_bytes key.g in
+    if Bn.(compare msg zero) <= 0 then raise (Error Negative_msg);
+    let gm = Bn.mod_power g msg n2 in
+    let rn = Bn.mod_power r n n2 in
+    let c = Bn.(mod_ (mult gm rn) n2) in
+    let c_bytes = Bn.to_bytes c in
+    Bn.wipe msg; Bn.wipe n; Bn.wipe n2; Bn.wipe g;
+    Bn.wipe r; Bn.wipe gm; Bn.wipe rn; Bn.wipe c;
+    c_bytes
+
+  let decrypt key c =
+    let c = Bn.of_bytes c in
+    let n = Bn.of_bytes key.n in
+    let n2 = Bn.of_bytes key.n2 in
+    let lambda = Bn.of_bytes key.lambda in
+    let mu = Bn.of_bytes key.mu in
+    let cn = Bn.mod_power c lambda n2 in
+    let lx = Bn.(div (sub cn one) n) in
+    let m = Bn.(mod_ (mult lx mu) n) in
+    let msg = Bn.to_bytes m in
+    Bn.wipe c; Bn.wipe n; Bn.wipe n2; Bn.wipe lambda;
+    Bn.wipe mu; Bn.wipe cn; Bn.wipe lx; Bn.wipe m;
+    msg
+
+  let new_key ?(rng = Random.secure_rng) numbits =
+    if numbits < 32 || numbits land 1 > 0 then raise(Error Wrong_key_size);
+    let numbits2 = numbits / 2 in
+    (* Make sure p > q *)
+    let rec gen_factors nbits =
+      let p = Bn.random_prime ~rng:(rng#random_bytes) nbits
+      and q = Bn.random_prime ~rng:(rng#random_bytes) nbits in
+      let cmp = Bn.compare p q in
+      if cmp = 0 then gen_factors nbits else
+      if Bn.(relative_prime (mult p q) (mult (sub p one) (sub q one))) then (p,q) else
+      (* Recursively generate factors *)
+      gen_factors nbits in
+    let (p, q) = gen_factors numbits2 in
+    (* n = pq *)
+    let n = Bn.mult p q in
+    let n2 = Bn.mult n n in
+    let g = Bn.add n Bn.one in
+    let lambda = Bn.(lcm (sub p one) (sub q one)) in
```

Comment:
Wikipedia (https://en.wikipedia.org/wiki/Paillier_cryptosystem) says that `lambda = (p - 1) * (q - 1)` is good enough, no need for a LCM.  But it's not obvious to me why, so I'm fine with the LCM.

Reply:
I guess its safe to find the LCM just incase